cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/FileFetcherHttpHandlerInternal.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.gateway.handlers;
+
+import com.google.common.io.Closeables;
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.io.Locations;
+import io.cdap.cdap.gateway.handlers.util.AbstractAppFabricHttpHandler;
+import io.cdap.cdap.proto.ApplicationDetail;
+import io.cdap.http.BodyProducer;
+import io.cdap.http.HttpHandler;
+import io.cdap.http.HttpResponder;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.http.EmptyHttpHeaders;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+
+import java.io.InputStream;
+import java.net.URI;
+import java.util.Base64;
+import javax.annotation.Nullable;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+
+
+/**
+ * Internal {@link HttpHandler} for serving file downloading requests.
+ *
+ * This is currently used by FileLocalizer to download required files from AppFabric
+ * in order to launch TwillApplication.
+ */
+@Singleton
+@Path(Constants.Gateway.INTERNAL_API_VERSION_3)
+public class FileFetcherHttpHandlerInternal extends AbstractAppFabricHttpHandler {
+  private static final Gson GSON = new Gson();
+
+  private final LocationFactory locationFactory;
+
+  @Inject
+  FileFetcherHttpHandlerInternal(CConfiguration configuration,
+                                 LocationFactory locationFactory) {
+    this.locationFactory = locationFactory;
+  }
+
+  /**
+   * Download the file specified by the base64-encoded URI
```

Comment:
Why base64 encode the URI? Shouldn't the file be always from the same base as the LocationFactory (i.e we shouldn't allow remote call to fetch files on the local fs of app-fabric).

Reply:
Without base64 encoding, the URL path becomes v3Internal/location/a/b/c/e/d, which results in not found, as file path gets treated as part of  the URL. 

Added a check in the http handler to ensure the file scheme is the same in the incoming URI and location factory.

Reply:
Base64 is still pretty unconventional in HTTP world. Can we do it as parameter and use regular URL encoding?

Reply:
You can use `**` to match any suffix. See https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeServiceRoutingHandler.java#L104 on how to use it.

Reply:
I was actually looking at it but was not sure you will get the whole thing into the last path variable

Reply:
Done. Fixed